firmware/library/L3_HAL/apds_9960.hpp

```diff
@@ -1,20 +1,518 @@
-  virtual void DecodeGestureData()                             = 0;
-  virtual uint8_t GetGesture()                                 = 0;
-  virtual uint8_t ReadMode()                                   = 0;
-  virtual uint8_t GetGestureEnterThreshhold()                  = 0;
-  virtual uint8_t GetGestureExitThreshhold()                   = 0;
-  virtual bool SetGestureEnterThresh(uint8_t threshhold_value) = 0;
-  virtual bool SetGestureExitThresh(uint8_t threshhold_value)  = 0;
-  virtual bool CheckIfGestureOccured()                         = 0;
-  virtual void ReadGestureFIFO()                               = 0;
+  enum Direction : char
+  {
+    kNone  = 'x',
+    kUp    = 'u',
+    kDown  = 'd',
+    kLeft  = 'l',
+    kRight = 'r',
+    kFar   = 'f',
+    kNear  = 'n'
+  };
+  enum Gesture : char
+  {
+    kSwipeUP    = 'U',
+    kSwipeDOWN  = 'D',
+    kSwipeLEFT  = 'L',
+    kSwipeRIGHT = 'R',
+    kNEAR       = 'N',
+    kFAR        = 'F',
+    kERROR      = 'E'
+  };
+  enum Mode : int8_t
+  {
+    kPOWERON = 0,
+    kALSENABLE,
+    kPROXIMITYDETECT,
+    kWAIT,
+    kALSINTERRUPT,
+    kPROXIMITYINTERRUPT,
+    kGESTURE
+  };
+
+  virtual bool FindDevice()                                       = 0;
+  virtual void Initialize()                                       = 0;
+  virtual bool SetMode(Mode mode, bool enable)                    = 0;
+  virtual void EnableGesture()                                    = 0;
+  virtual bool DisableGesture()                                   = 0;
+  virtual bool ReadGestureMode()                                  = 0;
+  virtual bool CheckIfGestureOccured()                            = 0;
+  virtual uint8_t GetGestureFIFOLevel()                           = 0;
+  virtual void ReadGestureFIFO(uint8_t * data, uint8_t fifolevel) = 0;
+  virtual void ProcessGestureData(uint8_t level)                  = 0;
+  virtual Gesture DecodeGestureData(uint8_t level)                = 0;
+  virtual Gesture GetGesture()                                    = 0;
+};
+
+class Apds9960 : public Apds9960Interface
+{
+ public:
+  // defualt values for sensor initialization
+  static constexpr uint8_t kNoMode = 0x00, kAtime = 0xDB, kWtime = 0xF6,
+                           kAiltl = 0xFF, kAilth = 0xFF, kAihtl = 0x00,
+                           kAihth = 0x00, kPilt = 0x00, kPiht = 0x32,
+                           kPers = 0x11, kConfig1 = 0x60, kPpulse = 0x89,
+                           kControl = 0x05, kConfig2 = 0x01,
+                           kPoffsetUpRight = 0x00, kPoffsetDownLeft = 0x00,
+                           kConfig3 = 0x00, kGPEnTh = 0x14, kGExTh = 0x64,
+                           kGconfig1 = 0x40, kGconfig2 = 0x66,
+                           kGoffsetUpDown = 0, kGoffsetLeftRight = 0,
+                           kGPulse = 0xC5, kGconfig3 = 0x00, kGconfig4 = 0x02;
+
+  bool FindDevice() override
+  {
+    i2c.Initialize();
+    uint8_t device_id = gesture.memory.device_id;
+    if (device_id == 0xAB)
```

Comment:
You should be able to do this in 1 line rather than having to store this in a variable first. You don't lose any efficiency, but it would look cleaner and the code would be shorter.